gensim/models/doc2vec.py

```diff
@@ -794,6 +794,24 @@ def train(self, documents=None, corpus_file=None, total_examples=None, total_wor
 
         """
         kwargs = {}
+
+        # Check if both documents and corpus_file are None
+        if corpus_file is None and documents is None:
+            raise TypeError("Either one of corpus_file or documents value must be provided.")
+
+        # Check if both documents and corpus_file are not None
+        if corpus_file is not None and documents is not None:
+            raise TypeError("Instead provide value to either of corpus_file or documents parameter but not both.")
+
```

Comment:
I'd be tempted to combine these two cases with an XOR: `if (corpus_file is None) ^ (documents is None):`.

Reply:
I did it this way for readability. Never mind, I'll make the changes.

Reply:
Yes, it'd actually need to be `if not ((corpus_file is None) ^ (documents is None)):` - but then essentially one error message, roughly "supply one or the other but not both", would be fine for either mistake.

Reply:
Strong -1 on this: unreadable. Please stick to simpler (and easier to maintain) constructs .

Reply:
The clearest approach would be not to have a method with such delicate XOR one-but-not-the-other positional parameters - use separate methods with distinct names! But failing that, a one-line XOR, whose next line is an error message with a plain-english description of the constraint, seems pretty simple to me - just accurately reflecting the design choice already made.